### PR TITLE
Add orderbook details to CLI and make a killer formatter

### LINKED ISSUE #166: Add Orderbook Details Endpoint to CLI
### Expected Behavior
```
stellar orderbook details <ID>
```

### Orderbook details to display
```
bids
asks
base
counter
```
https://www.stellar.org/developers/horizon/reference/endpoints/orderbook-details.html

### PR DESCRIPTION
**Pull request by correlator**
Resolves #166 

### Is there a GIF that reflects how this work made you feel?
Alt Text

### CODE REVIEW

File: cli/src/orderbook.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+use clap::ArgMatches;
+use error::Result;
+use fmt::{Formatter, Simple};
+use stellar_client::{endpoint::{orderbook, Limit}, resources::AssetIdentifier, sync::Client};
+
+pub fn details(client: &Client, matches: &ArgMatches) -> Result<()> {
+    let endpoint = {
+        let base_asset_type = matches
```
[the remark this answers is not in the document]
Comment (correlator) [reply cli/src/orderbook.rs]: a little `AssetIdentifier::from_matches` love sounds like a good time

File: cli/src/asset_identifier.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+use error::InvalidInputError;
+use stellar_client::resources::AssetIdentifier;
+
+pub fn parse_identifier_from_str(s: &str) -> Result<AssetIdentifier, InvalidInputError> {
+    let asset_parts: Vec<&str> = s.split("-").collect();
+    if asset_parts[0] == "xlm" {
```
Comment (kbacha) [inline cli/src/asset_identifier.rs]: Are identifiers case insensitive? Cause I think that lumens are "XLM".

File: cli/src/asset_identifier.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            "credit_alphanum4",
+            Some(asset_parts[0].to_string()),
+            Some(asset_parts[1].to_string()),
+        ).map_err(|e| InvalidInputError { details: e });
+    } else if asset_parts[0].len() < 13 {
+        return AssetIdentifier::new(
+            "credit_alphanum12",
+            Some(asset_parts[0].to_string()),
+            Some(asset_parts[1].to_string()),
+        ).map_err(|e| InvalidInputError { details: e });
```
Comment (kbacha) [inline cli/src/asset_identifier.rs]: have you tried `.into()`??

File: cli/src/error.rs [hunk 4]
```diff
@@ -18,7 +18,7 @@ pub enum CliError {
 /// Errors resulting from incomplete or invalid user input
 #[derive(Debug)]
 pub struct InvalidInputError {
-    details: String,
+    pub details: String,
```
Comment (kbacha) [inline cli/src/error.rs]: According to the `From` implementations I don't think you should have to expose this.

File: cli/src/asset_identifier.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
+use error::InvalidInputError;
+use stellar_client::resources::AssetIdentifier;
+
+pub fn parse_identifier_from_str(s: &str) -> Result<AssetIdentifier, InvalidInputError> {
+    let asset_parts: Vec<&str> = s.split("-").collect();
```
Comment (kbacha) [inline cli/src/asset_identifier.rs]: use `splitn` so that we don't split the issuer.

File: cli/src/asset_identifier.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    if asset_parts[0] == "xlm" {
+        return Ok(AssetIdentifier::Native);
+    }
+    if asset_parts.len() != 2 {
+        return Err(InvalidInputError {
+            details: "Asset identifier not of the form <asset_code>-<asset_issuer>".to_string(),
+        });
+    }
+    if asset_parts[0].len() < 5 {
+        return AssetIdentifier::new(
```
Comment (kbacha) [inline cli/src/asset_identifier.rs]: I think there are easier to use constructors than this now:

https://github.com/kbacha/stellar-sdk/blob/master/client/src/resources/asset.rs#L145-L163

File: cli/src/asset_identifier.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
+use error::InvalidInputError;
+use stellar_client::resources::AssetIdentifier;
+
+pub fn parse_identifier_from_str(s: &str) -> Result<AssetIdentifier, InvalidInputError> {
+    let asset_parts: Vec<&str> = s.split("-").collect();
+    if asset_parts[0] == "xlm" {
+        return Ok(AssetIdentifier::Native);
+    }
+    if asset_parts.len() != 2 {
```
Comment (kbacha) [inline cli/src/asset_identifier.rs]: You may try using pattern matching for this https://play.rust-lang.org/?gist=85d7266724a99db847d6741730ddebe6&version=stable&mode=debug

Comment (correlator) [reply cli/src/asset_identifier.rs]: *reacts with 🔥 emoji*

File: cli/src/asset_identifier.rs (hunk 5 above)
Comment (correlator) [reply cli/src/asset_identifier.rs]: Won't work if they put too many `-`'s in as splitn returns the rest of the string once it hit's the nth occurrence.  We could still check the last string to see if there is an extra `-` but no gain over current implementation and I like this one.  But thanks for pointing out splitn.  For sure good to know about.

File: cli/src/error.rs (hunk 4 above)
Comment (correlator) [reply cli/src/error.rs]: It keeps saying it's private.  Is there something I'm doing wrong or missing here?  If not, I'll write a from_str() method.

File: cli/src/main.rs [hunk 8]
```diff
@@ -283,40 +314,18 @@ fn build_app<'a, 'b>() -> App<'a, 'b> {
...
-                                    .takes_value(true)
-                                    .help("Filters trades with a counter_asset_code.  Not required for XLM"),
+                                    .required(true)
+                                    .help("Filters trades with a given base_asset. format:  <asset_code>-<asset_issuer>, or xlm if lumens"),
                             )
                             .arg(
-                                Arg::with_name("counter_asset_issuer")
-                                    .long("counter_asset_issuer")
+                                Arg::with_name("counter_asset")
+                                    .long("counter_asset")
```
Comment (kbacha) [inline cli/src/main.rs]: Do we really need to type `_asset` everytime we use this? I feel like just `base` and `counter` might be slightly easier to use on the command line.

File: cli/src/orderbook.rs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+pub fn details(client: &Client, matches: &ArgMatches) -> Result<()> {
+    let endpoint = {
+        let base_asset_str = matches
+            .value_of("base_asset")
+            .expect("Base asset is a required field");
+        let counter_asset_str = matches
+            .value_of("counter_asset")
+            .expect("Counter asset is a required field");
+        let base_asset = asset_identifier::parse_identifier_from_str(base_asset_str)?;
```
Comment (kbacha) [inline cli/src/orderbook.rs]: this is really long: `asset_identifier::parse_identifier_from_str`

I think we should consider that if we are using this from the module, the name of the method can probably just be `from_str`:

```rust
let base = asset_identifier::from_str(base_str)?;
```

File: cli/src/orderbook.rs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        let counter_asset = asset_identifier::parse_identifier_from_str(counter_asset_str)?;
+        let mut endpoint = orderbook::Details::for_asset_pair(base_asset, counter_asset);
+        if let Some(limit) = matches.value_of("limit") {
+            let limit = limit
+                .parse::<u32>()
+                .map_err(|_| String::from("Limitshould be a valid u32 integer"))?;
```
Comment (kbacha) [inline cli/src/orderbook.rs]: y'know, now that you have a constructor for the Invalid error type `from_str` you can probably use it for the return values.

**Pull request merged**